### ISSUE TITLE
Support for lane and turn lane tagging

### ISSUE THREAD

**Issue by simonpoole** [openstreetmap/iD#387]
The road style should reflect the number and type of lanes added. 

Further a smplified method of tagging lanes should be available.

See http://wiki.openstreetmap.org/wiki/Key:lanes and http://wiki.openstreetmap.org/wiki/Key:turn

**Comment by kepta** [comment]
Here is a (link)[https://docs.google.com/document/d/1HOM4TCKk2Bvw0tiRHQEh8B7gGrZqrQxBO0vH1bJidHw] to the proposal I submitted for my google summer of code application.

**Comment by slhh** [comment]
@kepta Thanks for sharing your proposal, I will comment later.
Do you know the awesome OSM Lane Visualizer?
Example: http://osm.mueschelsoft.de/cgi-bin/render.pl?relref=B%2075&start=1&placement&adjacent&lanewidth&usenodes
Supported Tags:https://github.com/mueschel/OsmLaneVisualizer#interpreted-tags
Maybe you can reuse some ideas, code or icons.

**Comment by slhh** [comment]
@kepta 
I don't think it is a good idea to make turn lanes unavailable somewhere for the following reasons:
- Turn lanes can have the length of multiple OSM-ways.If the user wants to tag the first section of such a turn lane first, his workflow will likely be prohibited by any rule for turn lane availability..
- Merge values can end without a junction.
- Special cases where other turn markings don't end at a junction do probably exist somewhere.

I don't see how splitting whithin the lane editor could work, because it needs to be made at the correct geoposition. I think the radial menu split operation is sufficient anyway.

What could be implemented in the lane editor is a longitudinal split due to a physical separation between lanes. It is a useful operation, but handling affected relations doesn't seem to be easy. Therefore, this has a low priority.

Instead of drawing allways rectangles the lanes should be drawn like following example:
![lane drawing](https://cloud.githubusercontent.com/assets/13684947/14868889/6cdd353e-0cd0-11e6-9966-5a16dee48c66.PNG)
The lane width should be drawn according to width:lanes, but at the bottom this should be overridden by width:lanes:start , and at the top be overridden by width:lanes:start.

The information defined using the keys 
placement, placement:start, placement:end altogether can be represented using a 4 vertix line like the red, dotted line in the example. Instead of these keys their :forward or :backward variants can be used alternatively. In oder to modify the placement keys I would like to just have the red, dotted line dragable.
In this case showing any of these placement keys as a field can be avoided.

It is important to support the witdth:lanes and placement keys including their :start and :end variants because they are not only required for lane rendering but also are very useful to define connectivity.
A well defined lane connectivity is absolutely necessary for lane support of navigation systems.

**Comment by slhh** [comment]
Users should not be required to use the lanes tagging syntax, therefore a lanes editor should not support a subset only of applicable keys. Therefore, I suggest a more generalized approch:
For each lane use a temporary object, both way lanes need one per direction. Decompose each `*:lanes*=*` tag of the highway into tags which are added to the lane objects, where `:lanes`, `:lanes:forward` and `:lanes:backward` is removed from the key and the value is reduced to the part corresponding to the respective lane.
The graphical UI of the lanes editor should allow the user to select a lane. When a lane is selected the "all fields" section and the "all tags" section should show and edit the tags of the temporary object of the selected lane instead of the highway object. This way the existing fields can be reused in order to edit the lane values. "All fields" and "all tags" should be temporarily named "lane fields" and "lane tags" respectively while the lane is selected.
When all lanes get unselected, the `*:lanes*=*` tag needs to be reassembled in order to be shown in the all tags section. There seems to be a preferred method to use :lanes for oneways and :lanes:forward/:lanes:backward otherwise, but in case this was tagged differently and the lane values haven't been changed, the original tag should be used.

**Comment by slhh** [comment]
The lane editor should not only cover keys with `:lanes`, `:lanes:forward`, and `:lanes:backward`, but also keys with pure `:forward` or `:backward` modifiers. Therefore, the previously mentioned approch should be extended:
Line the temporary lane objects use a temporary object for forward and backward direction each.
Tags of the highway object with `:forward` but not `:lanes:forward` key modifier go to the forward object with the modifier being cut out of the key. A thick forward arrow at the right side of the lanes drawing should enable selecting the forward direction. When selected the temporary forward object should be edited by all fields and all tags like a lane is edited according to my previous post.

The same applies with forward replaced by backward direction and the arrow at the left side of the lanes drawing.

**Comment by slhh** [comment]
The lane editor needs to care about right-hand or left-hand traffic, based on tagging of the highway and also the country-specific default. This is not only relevant for the graphics, but also for matching the value parts of  `*:lanes` tags to value parts of `*:lanes:forward` or `*:lanes:backward`.

**Comment by pnorman** [comment]
> but also for matching the value parts of `*:lanes` tags to value parts of `*:lanes:forward` or `*:lanes:backward`.

These are in the direction of the OSM way and I'm having a hard time coming up with a situation where this matters except for display. Could you explain?

Unfortunately, the wiki page is very short on examples of lane tagging for non-oneway roads.

**Comment by slhh** [comment]
Example with two forward and two backward lanes:
maxspeed:lanes=10|20|30|40 equals to

- maxspeed:lanes:backward=20|10 and maxspeed:lanes:forward=30|40 in case of right-hand traffic

- maxspeed:lanes:backward=40|30 and maxspeed:lanes:forward=10|20 in case of left-hand traffic

**Comment by HolgerJeromin** [comment]
Are you sure? The proposal says 
"In the common case of two driving directions we need to use the already existing :forward/:backward extension."

**Comment by slhh** [comment]
In the case of two driving directions using the pure `:lanes` extension might not be intended by the proposal, but the reality of the database seems to be different. In fact using the pure `:lanes` extension seems to be convenient for some base keys, and the meaning of the data is well defined.

**Comment by slhh** [comment]
@kepta 
Are you aware that the number of lanes defined by the `lanes*` keys doesn't need to equal to the number of lanes used in the values of keys with the `:lanes` suffix? Your proposal seem to assume that these quantities are equal.
The `lanes*` keys are counting full width lanes for motorized traffic only, but the `:lanes` suffix is counting all lanes without such a restriction.
Therefore, a common graphical representation of both `lanes*` and `*:lanes` doesn't seem to be possible.
I think the lane editor should focus on `*:lanes` and the `lanes*` keys should be supported by normal fields.
In case of missing `*:lanes` keys the quantities from `lanes*` keys can be used as a default for the graphical representation, but the user needs to be able to add lanes without changing the values of the `lanes*` keys.

**Comment by kepta** [comment]
Hey @slhh,  yes I am aware of that.
My mentor @bhousel did point that anomaly while I was writing the proposal
> 23:43 bhousel: also.. in some places, there is a special lane for a tram or something
23:44 bhousel: so it's not clear whether to tag that as a lane.. it's not a lane cars can drive in, but it's important to know about 
23:44 kepta_: hmm, I guess thats what the osm-talk discussion would be 
23:44 bhousel: yeah that's what they are discussing now
23:44 bhousel: but anyway, the weird thing is that the number in `lanes=` sometimes does not equal the sum of all the lanes
23:45 kepta_: :O
23:45 bhousel: yep, that's osm ¯\_(ツ)_/¯
23:45 kepta_: and why?
23:47 bhousel: because osm can't decide whether lanes means "lanes" or whether it means "lanes for motorized traffic"

I am sorry I forgot to mention it in my proposal.

**Comment by slhh** [comment]
@kepta, how do you intend to address that issue?

What bhousel did point out seems to be only a part of the problem.
It is not only the sum in `lanes=*` which might not equal to the sum of `lanes:forward=*`, `lanes:backward=*`, and `lanes:both_ways=*`, but also the value in for example `lanes:forward=*` doesn't need to match the count of lanes in `*:lanes:forward=*`
Example:
`bicycle:lanes:forward=no|designated|no|designated` usually requires `lanes:forward=2`.

> 23:45 kepta: and why?
> 23:47 bhousel: because osm can't decide whether lanes means "lanes" or whether it means "lanes for motorized traffic"

`lanes=*` is the older key, which seems to be mainly intended for capacity estimation, and its definition has been kept unchanged for compatibility reasons.
If the newer key suffix `:lanes` were restricted to full-width lanes for motorized traffic, describing some real situations like bicycle lanes between motorized traffic lanes would be prevented. 

Personally, I would prefer to deprecate `lanes=*`, and replace it with something based on `:lanes`, for example something like `direction:lanes=forward|forward|both_ways|backward`, which would also prevent any issues with driving_side (it is impossible to render lanes without determining driving_side) or even support any unusual sequence of driving directions.

**Comment by slhh** [comment]
@kepta, @bhousel
The intention of a lane editor can't be to let users spend time tagging useless data, but it should be to improve OSM data in order to enable applications. Therefore, the design should be driven by application requirements and not just by tag statistics.
Currently all the `*:lanes=*` tags are quite useless because existing applications using these tags are very rare.

I see the following main applications for the lane specific data, but these potential applications are prevented by missing data in most regions currently:

- **Rendering of lanes at high zoom levels** 
This needs the following keys for a harmonic position and shape:
`placement`, `placement:start`, `placement:end`, `width:lanes`, `width:lanes:start`, and `width:lanes:end`
The last two keys mainly support a more realistic shape of starting, ending, forking, or merging lanes.
In addition this application needs the following keys for road markings with second priority:
`turn`, `turn:lanes`, `change`, and `change:lanes`
- **Navigation system lane support**
This mainly needs information about lane connectivity. Connectivity of the lanes within a single way is defined by the `change` and `change:lanes` keys. Connectivity to the lanes of the following way can be defined by something like the `transit` key. I think the proposal should be improved, but we definitely need a lanes connectivity editor, which is working on interconnection nodes like the turn restriction editor, in order to define such connectivity. The lane connectivity editor might even be an extension of the restriction editor.
In simple cases such a connectivity specification is unnessesary or might be replaced by tagging `width:lanes:start`, and `width:lanes:end`, which allow to determine connectivity automatically in case of ending, starting, merging, or forking lanes. I think it is easier for the user to define the geometry of the single way than to define some kind of relation to the lanes of the next way.
Other lane properties seem to have second priority behind connectivity for this application.

The lanes editor should encourage users to provide the required information to enable these applications. In addition to all keys mentioned above. their `:forward` and `:backward` variants need to be supported too.

**Comment by slhh** [comment]
@kepta, @bhousel 
In addition to application requirements the priority of tag support of the lanes editor should be driven by programming efficiency and usability:

1. In order to avoid the requirement of major graphical redesigns the keys which need a special graphical representation and not just a an added icon should have first priority:
`width:lanes`, `width:lanes:start`, `width:lanes:end`, `placement`, `placement:start`, `placement:end`, `change`, `change:lanes`

2. Keys which would be error-prone to use if there were not a graphical representation should have second priority. The following keys do fall into this category if they are not already handled with first priority:
  - Due to potentially changing right/left with backward direction: `turn`, `turn:lanes`, `change`, `change:lanes`
  - Due to a potential confusion when combining `:start` or `:end` with backward direction:
`width:lanes`, `width:lanes:start`, `width:lanes:end`, `placement`, `placement:start`, `placement:end`

3. For the remaining tags with `:lanes` suffix a graphical representation is nice to have, but handling these tags with lane based fields seems to be sufficient for some time.

**Comment by HolgerJeromin** [comment]
Just for your information: OsmAnd uses turn:lane tagging for its navigation.

**Comment by kepta** [comment]
Here is an demo of what I have been working on.
Would love to have some suggestions.
![id_turn_lane](https://cloud.githubusercontent.com/assets/6966254/22148776/1f425ad2-df36-11e6-80fc-bfc8139bfb08.gif)

**Comment by HolgerJeromin** [comment]
Nice!
A graphical view of resulting arrows (most used combination are sufficient) above the lane number would be cool. 
So you could quick check (without clicking all tabs) if all lanes are matching the reality. 

Example of ways with both forward and backwards would be interesting, too.

**Comment by kepta** [comment]
@HolgerJeromin 
How about this 

![twoway](https://cloud.githubusercontent.com/assets/6966254/22628885/e7190812-ec01-11e6-89a9-0f29795cbb41.gif)

**Comment by HolgerJeromin** [comment]
I thought about adding it to the table in the sidebar. But in the map is perfect. 
Eventually they could be rendered even if the way is not selected (at a high zoom). This way checking a highway crossing would be extreme fast.

**Comment by slhh** [comment]
@kepta The lane icons on the map are awesome, but there is still some improvement possible:

- Use a pointing shape instead of the circular shape in order to indicate the lane direction. The current method is sufficiently clear in the example, but would become much less clear in case there aren't turn indications.
- Use a resizeable shape to allow placing multiple lane properties like turn and maxspeed.
- In case there are any hidden lane properties defined, which can't be visualized on the map, the style of the lane icon should indicate this.
- Highlight the lane icon of the selected lane.
- Make the lane selectable by clicking the lane icon on the map.

Example of pointing, resizeable lane icon shapes:
![lane icons](https://cloud.githubusercontent.com/assets/13684947/22630678/baf2e058-ebfe-11e6-818c-5250860d4417.jpg)

The arrows in the select lane field look like it were left-hand traffic, but the turn icons on the map are correct for right-hand traffic. The select lane field should have a "X" button to unselect all lanes and hide lane properties. Lane buttons of the select lane field should have a different styling in case any lane property has been set for the respective lane.

**Comment by tordans** [comment]
@kepta this looks awesome.

**Comment 1: Visualization:** 
I wonder why you went for the "icon on lane" approach instead of adding lanes like described on [OSM Lane Visualizer](http://osm.mueschelsoft.de/cgi-bin/render.pl?relref=B%2075&start=1&placement&adjacent&lanewidth&usenodes). I guess its a lot harder to create those lines on the fly?


**Comment 2: Editor Usability:** 
I think the UI you proposed at https://github.com/openstreetmap/iD/issues/387#issuecomment-277539927 could be improved. Especially when considering the interface design patterns that the iD Editor uses.

The design pattern – the way I understand it – creates a group ("Lanes") and in this group allows for key-value-editing. The editing itself is based on dropdowns most of the time. 

Sitenote: It also uses radio button lists only in very few cases and in your wireframe 1 it should be at least a radio button list – but the dropdown approach reduces the overall visual complexity, so its even better from a usability point of few.

Based on this, the UI could be [(URL)](https://cyakew.axshare.com/#c=2):
<img width="850" alt="bildschirmfoto 2017-12-30 um 07 53 12" src="https://user-images.githubusercontent.com/111561/34452011-88312ba4-ed36-11e7-81ec-134bfbcb2e80.png">

Your design is based on indenting the panels but they are still separate panels that have an interaction between each other. This is not what I would expect them to behave like and I think it is quite komplex and therefore hard to understand and use. This might be a good place for some quick user testing with 5 mappers ("new" and "old" once).

My wireframe 2 follows the existing patterns more closely. It has however a disadvantage for power users that want to edit with as few clicks as possible (editing speed). However, this is an advantage in my opinion since the iD editor targets new users over power user (who have tools like JOSM).

Possible extentions:
- The (i) icon should preview and explain the edit with an example image just like it does in other cases
- Since the order of tags from top to bottom has a meaning it might be a good idea make this more explicit by adding a number "1. Lane", "2. Lane" as a prefix to the row.
    <img width="412" alt="bildschirmfoto 2017-12-30 um 08 05 45" src="https://user-images.githubusercontent.com/111561/34452114-5cdfe380-ed38-11e7-8991-b994a79506c0.png">


**Comment 3: Bike lanes:** 
Did you look at https://wiki.openstreetmap.org/wiki/Key:cycleway for this project and do you have plans to add a proper bike lane editing to the Editor? For example, there is the whole [left/right vs. forward/backward](https://wiki.openstreetmap.org/wiki/Forward_%26_backward,_left_%26_right) complexity to handle.

**Comment by quincylvania** [comment]
I felt inspired today and mocked-up the basics of a potential lane editor. It has icons for the turning and traffic type of each lane. You would click the icons to cycle through the options if there are only a couple, or open a picker if there are lots. You would drag-and-drop the lanes or dividers to change the arrangement. I'm not quite sure how adding/deleting lanes would work yet.

I hadn't read this thread beforehand so there are probably tips here I could incorporate. Good idea to show the cross-streets!

![lanes 1](https://user-images.githubusercontent.com/2046746/49979993-5e31aa80-ff06-11e8-8513-94cbe79bad19.png)
![lanes 2](https://user-images.githubusercontent.com/2046746/49979994-5e31aa80-ff06-11e8-8a09-bab2699e7de6.png)

**Comment by bhousel** [comment]
Looks awesome @quincylvania 👍

**Comment by slhh** [comment]
@quincylvania
Looks nice, but we do also need an identification of lane position and direction on the map.
When hovering a lane in the sidebar the rendering of the selected highway on the main map can change to rendering separate lines per lane including direction arrows per lane, and highlighting of the hovered lane.

To add or delete lanes the lane representation in the sidebar can have a  right click context menu  with delete, add on right, and add on left operations.

Showing cross-streets looks nice at a glace, but I think it would be less useful or even confusing in practice: It does likely work for small roads, but these rarely need lane tagging. Larger roads with multiple lanes do unlikely have constant properties between two cross streets. This results in splitted OSM ways between cross streets. 
Another issue is the naming of highways within large junctions, e.g. with separate ways per direction.

**Comment by bhousel** [comment]
tbh, the more I think about this, the more I'd prefer to just render the lanes directly on the actual map at high zooms.

**Comment by 1ec5** [comment]
Rendering the lanes onto the map would address the issue of orienting the user, at least when they aren’t copying from street-level imagery. Would we have the user edit the lanes directly on the map, or would there still be lane-specific controls in the sidebar? As soon as we put anything in the sidebar, we have to address the issue of associating the map overlays with sidebar controls, which is challenging when the way runs south to north or non-linearly.

Perhaps the user could select a lane on the map, and the sidebar would show only the controls for the selected lane, as though it were a first-class object like a node or way. Not sure where the user would go to add a lane.

Lane overlays like the ones shown in https://github.com/openstreetmap/iD/issues/387#issuecomment-277539927 make plenty of sense for `turn:lanes`, but if we ever want to support `placement` or `change:lanes`, rendering those details as overlays would be tricky. JOSM takes the approach of completely obscuring the road with a lane diagram that can depict all those details, but while that approach great for understanding what’s already tagged, I think obscuring any amount of aerial imagery would be inconvenient for anyone mapping turn lanes from scratch. Also, I wonder if people would try to drag these overlays around, since they wouldn’t look like a contextual menu.

I tend to edit at zoom level 18 or 19, because I have the luxury of crisp aerial imagery at those zoom levels where I map. I’d probably find it annoying to have to zoom in further to map turn lanes, since I’d have to zoom back out to easily pan to the next intersection. For reference, this is what a wide, two-way road with lots of turn lanes looks like at z19 around 39°N:

<img width="446" alt="kemper" src="https://user-images.githubusercontent.com/1231218/50086876-521e4500-01b3-11e9-86fa-8f188dfe3a6e.png">

This is a more common scenario, where I would typically enable wireframe mode to see what’s under the rendered way:

<img width="488" alt="adams" src="https://user-images.githubusercontent.com/1231218/50087508-200de280-01b5-11e9-80a9-e312f625894e.png">

It sounds like you’d only show lanes at a higher zoom level where there’s more screen real estate, to avoid overlapping adjacent features like sidewalks. But a higher zoom level could still get uncomfortable if the user happens to be using an imagery layer that gets unreadably pixelated above z19.

**Comment by jguthrie100** [comment]
It could always be an option to orientate the lane editor in the same direction as the map and have it rotate around as the map is rotated..

**Comment by ignaciolep** [comment]
Please, bear in mind the need to input different maxspeed values for each lane.
This is important in two-way road, where each direction have different maxspeed values.
Wiki: https://wiki.openstreetmap.org/wiki/Key:maxspeed#Lanes

**Comment by BjornRasmussen** [comment]
> tbh, the more I think about this, the more I'd prefer to just render the lanes directly on the actual map at high zooms.  

That could be setting that is automatically activated when editing lanes.

**Comment by BjornRasmussen** [comment]
Lanes would only be rendered when they are needed.

**Comment by quincylvania** [comment]
Here's an update on my previous design. You would click on a lane to open an editing popover.

<img width="388" alt="Screen Shot 2019-04-15 at 7 54 12 AM" src="https://user-images.githubusercontent.com/2046746/56142696-f773da80-5f53-11e9-8145-e8361e371d31.png">

As noted elsewhere, special lanes like sidewalks, bike lanes, and parking lanes should all be incorporated into the lane editor.

It'd be great to show the lane editor directly on the map in some cases, but I think it's necessary to always have it in the sidebar too.

**Comment by 1ec5** [comment]
I don’t think we can completely rely on cross street names to disambiguate directions. For example, [this way](https://www.openstreetmap.org/way/68507299) doesn’t have any named cross streets, its endpoints are continuations of the current street, and the nearest named cross streets at each end are the same street. If I were to add street parking along one side but not the other, I’m not sure which side in the diagram I’d edit. The cross street name can be an aid, but there probably needs to be something else too.

In this latest design, the diagram wouldn’t be directly editable, so I wonder if we could get away with rotating it according to the main map, just as with the turn restriction mini-editor.

**Comment by quincylvania** [comment]
> I don’t think we can completely rely on cross street names to disambiguate directions.

Agreed. We would only show the crossroad like that if it was a clear circumstance.

> In this latest design, the diagram wouldn’t be directly editable, so I wonder if we could get away with rotating it according to the main map, just as with the turn restriction mini-editor.

Possibly. I'm not sure how this would work in practice. I'd prefer to just rotate the map to align with the editor, but this wouldn't necessarily work especially well with long and twisty roads. We could try larger way direction markers when editing roads, or even an animation on the road showing the flow.

**Comment by slhh** [comment]
@quincylvania 
When a lane is hovered or selected in the sidebar , dots moving along the lane could be animated on the main map to indicate lane direction. Preferably, the lane position is calculated based on the `driving_side`, `placement`, and `width:lanes` tags where present. The slight side ofset to the OSM way would be visible on high zoom levels only, but the lane direction on all editing zoom levels.

**Comment by slhh** [comment]
@quincylvania 
>I'd prefer to just rotate the map to align with the editor,

The rotation might not be identifyable for the user dependent on the structure of the data and thebackground structure in the viewport. The user might loose orientation.

>We would only show the crossroad like that if it was a clear circumstance.

What seems to be a clear circumstance for Id might be different what the user thinks to be clear. iD might show the crossroad on the other side of the way than expected by the user. This could result in inverse tagging of the lanes.

Showing crossroads could be quite confusing within large junctions with separate carriageways for each directing and/or separate carriageways for turning. Especially the later ones rarely have well defined names.

**Comment by BjornRasmussen** [comment]
When I first started editing osm with iD, the bike lane editor had the right and left side editor, which was confusing at first.   I glanced at the selected way, saw the small arrows, and mapped according to which side was left and right.  I don't think that this should be the number one concern

**Comment by BjornRasmussen** [comment]
One simple solution would be rendering the cardinal directions in the editor as subtle letters on the sides.  This would let people know that there is a directionality to the editor.

**Comment by BjornRasmussen** [comment]
I have a simple UI suggestion: make the editor have two sections, a visualizer that can let you select lanes and lane dividers, and an editor for specifying lane information for the selected lane/divider.  
Basically just a more beautiful version of this:    
![image](https://user-images.githubusercontent.com/37753777/56220197-0b900880-6036-11e9-9369-cf2c0d0355eb.png)
    
What I mean by "divider" is that you could click on a divider and specify change:lanes for nearby lanes, and therefore the type of divider (dashed vs solid).

Lane width would also be useful, especially since bike lanes are usually much narrower than drive lanes.

Also, the bottom area could be used for editing the number of lanes when no lane is selected.  This way, no space is wasted.

**Comment by BjornRasmussen** [comment]
Another idea I have is render the lane display based on a .json rendering file that can be different depending on which country the lanes are being edited in.  Most of Europe does not have a double yellow stripe in the center of the road, but rather a double white stripe.

**Comment by BjornRasmussen** [comment]
Of course, a popover with the same information as my sketch would likely be better, since even less space is used.

**Comment by slhh** [comment]
@BjornRasmussen 
Cardinal directions wouldn't work for U-shaped or circular ways.

When the sidebar is open, using a popover requires even more space. In addition, a popover needs a control to be opened.
Users need to be made aware of any existing lane tagging, major geometric changes like merging or appending a way are otherwise prone to generate wrong data. Therefore, we must not hide lane tagging in a rarely opened popover.

**Comment by slhh** [comment]
@quincylvania 
>Here's an update on my previous design. 

You need to differentiate pure forward/backward indications and "through" turn indications.

**Comment by tordans** [comment]
Hi! I am very happy to see new energy in this thread.

However, I don't know if I still know what the goal of this thread is about.

I write this, because I see a real issue with trying to solve all lane related tagging in one place. Which will IMO become a huge mess of complexity (like this thread show already). And I am really not sure its even possible to build a UI for all this.

In case we try to find a UI that "consolidate(s) all of the "stuff tagged on a street" into a single place" ([like bhousel put it](https://github.com/openstreetmap/iD/issues/6178#issuecomment-482986614) and [quincylvania visualizes it](https://github.com/openstreetmap/iD/issues/387#issuecomment-483289342)), I see two mayor issues:

**1. Separate lanes vs. "everything on one lane".**

Most mockups above assume IMO, that there is one way which holds all the tags. Eg https://github.com/openstreetmap/iD/issues/387#issuecomment-483702193 and https://github.com/openstreetmap/iD/issues/387#issuecomment-483289342.

The advantage is, that the UI looks very good and easy to use.
The disadvantage is, that its only one way of tagging and it does not really work if lanes are split in multiple ways or other means of transport (bike, foot) are split in separate ways.

I also feel like the community is slowly pushing towards tagging separate was for separate traffic sources. Like https://www.youtube.com/watch?v=DHecYP_b3os (a good talk!). This development is IMO accelerated by the need for more descriptive tagging (like bike-lane-width and -smoothness, …) on ways, which will either cause the main-way to be segmented into a hundred peaces (making it impossible to maintain) or it will push towards separating eg. the means of transport.

**2. Not just lane tags, but all the sub-tags as well**

The other issue is, that having a UI that handles the lanes is one thing. But the same UI should also allow to add all the descriptive sub-tags as well. I see https://github.com/openstreetmap/iD/issues/387#issuecomment-483702193 started adding a few of those. But there a quite a few more. And they are different per "lane" (eg. parking-sub-tags vs. driving-sub-tags). This will make the UI very complex.

### Two approaches

**One UI for all**

Lets assume we find a UI that can deal with the separate lane-issue (1) above. Which might be possible given enough dev-power! I feel like the discussion here is not really clear about what we want to add to this UI. What are the features we need to add to this UI? IMO https://streetmix.net/-/773938 is a good start:
- sidewalks
- parking (car, bike, …)
- car driving
- bike driving
- bus driving
- tram (train) on street
- tram (train) as separate lane (eg. in the middle, between car lanes)
- and between all of those separation elements like …
  - barrier=kerb
  - trees
  - bollards

If this thread continues this road, I suggest we first collect the cases we want to handle with the UI. The list above could be a starting point.

**Separate UIs**

On the other hand, we could just start by adding a great UI per use-case.

- for car driving there is a lot of good stuff above, that will make the IMO very hard to understand tagging schema of lanes really easy to use. This is a huge win, IMO. Just like turn lane-UI made a very complex tagging super easy; by focusing on this one problem.
- for parking, I started https://github.com/openstreetmap/iD/issues/6178 to collect ideas
- for bike lanes – I fear there is a lot to do on the tagging definition front, first
- bus, tram, … – I personally know to little about, yet

Maybe we could even think about how to push the mapping forward in the direction of "separate lanes per means of transport". Eg by adding a "split the sidewalk (or cycle lane) tags from this way into a separate way"-feature.

**Comment by quincylvania** [comment]
@BjornRasmussen said:

> I have a simple UI suggestion: make the editor have two sections, a visualizer that can let you select lanes and lane dividers, and an editor for specifying lane information for the selected lane/divider.

Yep, I was thinking something like this as well. Either below the lanes, or as I said in my one comment, "You would click on a lane to open an editing popover." This could also apply to dividers if the OSM tagging is that far developed. 

@tordans said:

> I also feel like the community is slowly pushing towards tagging separate was for separate traffic sources.

I'm all for this! I vastly prefer separate sidewalks and even bike lanes, but if people keep tagging this stuff on roads then I do think it makes sense to show them in the lane layout.

> The other issue is, that having a UI that handles the lanes is one thing. But the same UI should also allow to add all the descriptive sub-tags as well.

We can start with the basic lane info (access, turns) and see what makes sense after that. The point is to make lane-tagging possible and easy, not to do everything 😅

**Comment by tordans** [comment]
> The point is to make lane-tagging possible and easy, not to do everything 😅

I am glad to hear that. 

That would mean, we can now find **the best solution to handle the following tags**, right?

- `lanes=3…`
- `turn:lanes=none|none|none|merge_to_left|…`

and
- `lanes:forward=1…`
- `turn:lanes:forward=left|through;right…`
- `lanes:backward=1…`
- `turn:lanes:backward=through|through;right…`

Source: https://wiki.openstreetmap.org/wiki/Key:turn#Example_for_a_road_with_both_directions

But also destination tags? - https://wiki.openstreetmap.org/wiki/Key:destination
Anything else?

**So we need to be able to …**

- add a lane
- remove a lane
- reorder lanes
- specify the lane type
- change the direction of a lane
- respect the oneway-key
- understand where north is / which way the street shows / what direction I am editing
- anything else?

**If so, the first question for the UI is IMO:**

- should it display the lanes on the road (in the map view) (https://github.com/openstreetmap/iD/issues/387#issuecomment-447620208 and maybe https://github.com/openstreetmap/iD/issues/387#issuecomment-277539927)

- or: should it handle the details in the sidebar (https://github.com/openstreetmap/iD/issues/387#issuecomment-483702193 and similar)

- or: both, but the map to show it, the sidebar to edit it?

**Comment by BjornRasmussen** [comment]
> should it display the lanes on the road (in the map view) (#387 (comment) and maybe #387 (comment)). 

> or: should it handle the details in the sidebar (#387 (comment) and similar). 

> or: both, but the map to show it, the sidebar to edit it?  
  
I would prefer both, with the option to turn renewing of lanes situated next to the wireframe option in `map data`.

**Comment by BjornRasmussen** [comment]
> But also destination tags? - https://wiki.openstreetmap.org/wiki/Key:destination
  
I think that it would be too difficult to work this into the interface.

**Comment by EmpireFall** [comment]
I was designating 'turn:lane' tags for 'ways' using the current iD editor, it requires a certain amount of focus to get right, otherwise the user may confuse the direction of the 'way' with the intended direction of the 'tag' in conjunction with the road markings seen in the background layer (as the forward heading direction of a 'way' is defined by the direction in which it is drawn, there have been times that I have mistakenly assumed that the forward heading direction of the way is the same as the side of the road in which traffic flows, which has led me to designate tags accordingly only to realize later that the heading of the way is in reverse, that coupled with a background layer that is always facing upright can make designating turn:lane tags a rather focus intensive processs)

in this regard, a function in the iD editor that came to mind, that I thought was rather intuitive is the 'turn restriction' box 

would it not be possible to present the 'turn:lane', 'parking:lane', 'road markings, etc' tags for each respective lane with similar presentation and functionality of the 'turn restriction' box?

if the angle for each symbol of each respective lane tag were facing the same orientation as the 'way' itself (as seen in the turn restriction box), that would make it much easier for the user to orient themselves with the respective tags of each lane in conjunction to the orientation of the 'background layer'

https://i.imgur.com/PJxpPip.jpg
https://i.imgur.com/96k12Rj.jpg

as the orientation in which a 'way' can be either 'forward' or 'backward' and angled toward any direction, having a box with all the lane details presented facing upwards (as shown in the image below), the user may end up feeling as disoriented as with the current setup available in the iD editor

https://i.imgur.com/vhNHga0.png

imagine designating 'turn restrictions' if the orientation of the 'way' in the turn restriction box were facing upright rather than at the orientation of the way itself, things could get rather confusing without a certain amount of focus

just a thought

**Comment by hungerburg** [comment]
Removing the WIP label usually means, the feature is complete. Can I get my hands on a working testbed?

**Comment by kymckay** [comment]
@hungerburg Removal of WIP just signifies work is no longer actively in progress. There'd be a lot more buzz on this issue if the feature was complete 😄

**Comment by hungerburg** [comment]
Is the abandoned work of Quincy's open source? Recently someone did some lane tagging in the area of my local_knowledge with iD and I think, they mostly did that manually, at least not all the stuff passes validation, so I think that it be worthwhile to continue that work, the screenshots looking very nice indeed.

**Comment by machineonamission** [comment]
bumping this issue because there's multiple designs and working prototypes and I really want to see this integrated into iD.  JOSM's turn lane plugin sucks

**Comment by Zaczero** [comment]
@HexCodeFFF, I think that's a trend with iD development in general. I wouldn't get my hopes up. I completely gave up on any iD contributions a few weeks after posting this thread: https://github.com/openstreetmap/iD/issues/9500

**Comment by machineonamission** [comment]
Damn. Sucks cause iD is, imo, much cleaner and easier to use than JOSM but there's no plugins and slow development.

**Comment by dabreegster** [comment]
Putting in a plug for https://github.com/a-b-street/osm2streets/issues/216 -- we have a web app for visual lane tagging started. Could use help with getting OAuth to work, and with design/implementation for a nicer UI to edit (rather than raw tags). Happy to adapt/use one of the designs from earlier in this thread -- better to get something going and iterate quickly.
![Screenshot from 2023-05-30 23-31-36](https://github.com/openstreetmap/iD/assets/1664407/8f04306d-58c6-44c1-8f91-76f52d4792df)
